akka-discovery-azure-api/src/main/scala/akka/discovery/azureapi/AzureRbacAKSServiceDiscovery.scala

```diff
@@ -0,0 +1,270 @@
+
+  /**
+   * INTERNAL API
+   *
+   * Finds relevant targets given a pod list. Note that this doesn't filter by name as it is the job of the selector
+   * to do that.
+   */
+  @InternalApi
+  private[azureapi] def targets(
+      podList: PodList,
+      portName: Option[String],
+      podNamespace: String,
+      podDomain: String,
+      rawIp: Boolean,
+      containerName: Option[String]): immutable.Seq[ResolvedTarget] =
+    for {
+      item <- podList.items
+      if item.metadata.flatMap(_.deletionTimestamp).isEmpty
+      itemSpec <- item.spec.toSeq
+      itemStatus <- item.status.toSeq
+      if itemStatus.phase.contains("Running")
+      if containerName.forall(name =>
+        itemStatus.containerStatuses match {
+          case Some(statuses) => statuses.filter(_.name == name).exists(!_.state.contains("waiting"))
+          case None           => false
+        })
+      ip <- itemStatus.podIP.toSeq
+      // Maybe port is an Option of a port, and will be None if no portName was requested
+      maybePort <- portName match {
+        case None =>
+          List(None)
+        case Some(name) =>
+          for {
+            container <- itemSpec.containers
+            ports <- container.ports.toSeq
+            port <- ports
+            if port.name.contains(name)
+          } yield Some(port.containerPort)
+      }
+    } yield {
+      val hostOrIp = if (rawIp) ip else s"${ip.replace('.', '-')}.$podNamespace.pod.$podDomain"
+      ResolvedTarget(
+        host = hostOrIp,
+        port = maybePort,
+        address = Some(InetAddress.getByName(ip))
+      )
+    }
+
+  final class KubernetesApiException(msg: String) extends RuntimeException(msg) with NoStackTrace
+
+  final case class KubernetesSetup(namespace: String, ctx: HttpsConnectionContext)
+}
+
+class AzureRbacAKSServiceDiscovery(implicit system: ActorSystem) extends ServiceDiscovery {
+  private val http = Http()
+
+  private val settings = Settings(system)
+
+  private val log = Logging(system, classOf[AzureRbacAKSServiceDiscovery])
+
+  log.debug("Settings {}", settings)
+
+  private def fetchToken: Future[AccessToken] =
+    AzureRbacAKSServiceDiscovery.azureDefaultCredential
+      .getToken(AzureRbacAKSServiceDiscovery.accessTokenRequestContext.addScopes(settings.entraServerId))
+      .toFuture
+      .asScala
+
+  private val kubernetesSetup = {
+    implicit val blockingDispatcher: ExecutionContext = system.dispatchers.lookup(DefaultBlockingDispatcherId)
+    for {
+      namespace: String <- Future {
+        settings.podNamespace
+          .orElse(readConfigVarFromFilesystem(settings.podNamespacePath, "pod-namespace"))
+          .getOrElse("default")
+      }
+      httpsContext <- Future(clientHttpsConnectionContext())
+    } yield {
+      KubernetesSetup(namespace, httpsContext)
+    }
+  }
+
+  import system.dispatcher
+
+  private def accessToken(retries: Int = 3): Future[AccessToken] = {
```

Comment:
There is nothing wrong with the if/else but when writing scala I 
```suggestion
  private def accessToken(retries: Int = 3): Future[AccessToken] = {
  fetchToken.flatMap {
    case token if token.isExpired && retries > 0 => accessToken(retries - 1)
    case token if token.isExpired => Future.failed(new RuntimeException("Failed to fetch a valid token after multiple attempts"))
    case token => Future.successful(token)
  }
}
```

Reply:
Agreed. I changed it from if/else